### PR TITLE
Omar Bassyouni - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by Omar-Bassyouni**

### CODE REVIEW

File: MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 1]
```diff
@@ -280,6 +280,29 @@ final class FeedUIIntegrationTests: XCTestCase {
 		wait(for: [exp], timeout: 1.0)
 	}
 
+	func test_loadFeedCompletion_rendersErrorMessageOnErrorUntilNextReload() {
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		loader.completeFeedLoadingWithError()
```
Comment (danillahtin) [inline MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift:287]: It should be tested that errorView is not visible when the view is loaded
```suggestion
		sut.loadViewIfNeeded()
		XCTAssertEqual(sut.isErrorVisible, false)
		loader.completeFeedLoadingWithError()
```

File: MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 2]
```diff
@@ -280,6 +280,29 @@ final class FeedUIIntegrationTests: XCTestCase {
 		wait(for: [exp], timeout: 1.0)
 	}
 
+	func test_loadFeedCompletion_rendersErrorMessageOnErrorUntilNextReload() {
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		loader.completeFeedLoadingWithError()
```
Comment (danillahtin) [inline MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift:287]: It should be tested that errorView is not visible when the view is loaded
```suggestion
		sut.loadViewIfNeeded()
		XCTAssertEqual(sut.isErrorVisible, false)
		loader.completeFeedLoadingWithError()
```

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 3]
```diff
@@ -18,16 +18,29 @@ final class FeedViewModel {
 		Localized.Feed.title
 	}
 
-	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onStateChange: Observer<State>?
```
Comment (danillahtin) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:22]: The concept of `loading` and `error` should be separate since `errors` are not just for loading errors. We can represent errors for many other operations.

So ideally, there should be a separate observer for the error, as shown in the lectures.

```suggestion
	var onLoadingStateChange: Observer<Bool>?
	var onErrorStateChange: Observer<Bool>?
```

This way, you could have added this new feature without changing any code related to `loading`!

**Review by danillahtin — COMMENTED** [review]
Hi @Omar-Bassyouni, you almost passed the challenge!

This is Danil, I am an instructor at Essential Developer.

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

**Comment by Omar-Bassyouni** [comment]
> Hi @Omar-Bassyouni, you almost passed the challenge!
> 
> This is Danil, I am an instructor at Essential Developer.
> 
> Please, check my review comments and let me know once you're done so I can review it again.
> 
> Thank you 🙂

Hi @danillahtin,

I have done the suggested improvements, I just have one question which is a replay for a comment regarding having errorStateObsevable & loadingStateObservable rather than one stateObservable.  please take a look :)

Thank you very much for your feedback, really helpful and beneficial.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift (hunk 3 above)
Comment (Omar-Bassyouni) [reply MVVM Module/Source/Feed Presentation/FeedViewModel.swift]: this approach was inspired by mentoring session #008 for MVVM and RxSwift. 
I remember Caio saying it makes it much easier to combine streams into one stream, otherwise it gets sometimes out of hand.
In this session, there were no loading and errors but only the state of the form, maybe things that are related are advisable to be in one stream but things like loading and error where they are not related are in different streams. 
I am not sure whats your opinion is on this?

Comment (caiozullo) [reply MVVM Module/Source/Feed Presentation/FeedViewModel.swift]: Hi @Omar-Bassyouni! In session 008, I recommended keeping the *table state* (sections/rows state) in a single property. But unrelated functionality should be separated in other streams/observers because unrelated parts shouldn't affect others. For example, changing the error state shouldn't alter the table sections/rows - so they should be separate. In the same way, loading and error states are separate concerns. Nothing stops us from showing an error while still loading (and vice-versa!).

**Review by danillahtin — COMMENTED** [review]
Hi @Omar-Bassyouni,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

Great job!

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift (hunk 3 above)
Comment (Omar-Bassyouni) [reply MVVM Module/Source/Feed Presentation/FeedViewModel.swift]: ok, got it now.
Awesome thanks for the amazing PR 🚀
